modules/core/src/main/java/org/apache/ignite/internal/QueryMXBeanImpl.java

```diff
@@ -0,0 +1,168 @@
+import org.apache.ignite.internal.visor.query.VisorQueryCancelTaskArg;
+import org.apache.ignite.internal.visor.query.VisorScanQueryCancelTask;
+import org.apache.ignite.internal.visor.query.VisorScanQueryCancelTaskArg;
+import org.apache.ignite.mxbean.QueryMXBean;
+
+import static org.apache.ignite.internal.sql.command.SqlKillQueryCommand.parseGlobalQueryId;
+
+/**
+ * QueryMXBean implementation.
+ */
+public class QueryMXBeanImpl implements QueryMXBean {
+    /** */
+    public static final String EXPECTED_GLOBAL_QRY_ID_FORMAT = "Global query id should have format " +
+        "'{node_id}_{query_id}', e.g. '6fa749ee-7cf8-4635-be10-36a1c75267a7_54321'";
+
+    /** */
+    private final GridKernalContext ctx;
+
+    /** */
+    private final IgniteLogger log;
+
+    /**
+     * @param ctx Context.
+     */
+    public QueryMXBeanImpl(GridKernalContext ctx) {
+        this.ctx = ctx;
+        this.log = ctx.log(QueryMXBeanImpl.class);
+    }
+
+    /** {@inheritDoc} */
+    @Override public void cancelContinuous(String routineId) {
+        A.notNull(routineId, "routineId");
+
+        cancelContinuous(UUID.fromString(routineId));
+    }
+
+    /** {@inheritDoc} */
+    @Override public void cancelSQL(String id) {
+        A.notNull(id, "id");
+
+        if (log.isInfoEnabled())
+            log.info("Killing sql query[id=" + id + ']');
+
+        boolean res;
+
+        try {
+            IgniteClusterImpl cluster = ctx.cluster().get();
+
+            T2<UUID, Long> ids = parseGlobalQueryId(id);
+
+            if (ids == null)
+                throw new IllegalArgumentException("Expected global query id. " + EXPECTED_GLOBAL_QRY_ID_FORMAT);
+
+            res = cluster.compute().execute(new VisorQueryCancelTask(),
+                new VisorTaskArgument<>(ids.get1(), new VisorQueryCancelTaskArg(ids.get1(), ids.get2()), false));
+        }
+        catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+
+        if (!res)
+            throw new RuntimeException("Query not found[id=" + id + ']');
+    }
+
+    /** {@inheritDoc} */
+    @Override public void cancelScan(String originNodeId, String cacheName, Long id) {
+        A.notNullOrEmpty(originNodeId, "originNodeId");
+        A.notNullOrEmpty(cacheName, "cacheName");
+        A.notNull(id, "id");
+
+        if (log.isInfoEnabled())
+            log.info("Killing scan query[id=" + id + ",originNodeId=" + originNodeId + ']');
+
+        cancelScan(UUID.fromString(originNodeId), cacheName, id);
+    }
+
+    /**
+     * Kills scan query by the identifiers.
+     *
+     * @param originNodeId Originating node id.
+     * @param cacheName Cache name.
+     * @param id Scan query id.
+     */
+    public void cancelScan(UUID originNodeId, String cacheName, long id) {
+        boolean res;
+
+        try {
+            IgniteClusterImpl cluster = ctx.cluster().get();
+
+            ClusterNode srv = U.randomServerNode(ctx);
```

Comment:
Why do we use random node here? IMO we need to send task to the origin node and run `GridCacheDistributedQueryFuture#cancelQuery` there.

Reply:
Same of previous:

> Why it's not a good idea to send cancel message to the initiator

My opinion - canceling of the query is the "last line of defense".
When an administrator decides to cancel a query it means something goes wrong.
The whole initiator node can be flooded and respond slowly.
So I think it's not the best choice to send a cancel request to the initiator.

> Why random node?

Actually, scan query cancel executed on all cluster nodes it implemented in the task itself see VisorScanQueryCancelTask#jobNodes:
```
    /** {@inheritDoc} */
    @Override protected Collection<UUID> jobNodes(VisorTaskArgument<VisorScanQueryCancelTaskArg> arg) {
        return F.transform(ignite.cluster().nodes(), ClusterNode::id);
    }
```

Reply:
I think we need to have a single entry point for the query cancellation. And this entry point should be located on the initiator. I can name a few reasons for it:

1. Convenience: single entry point is easier to control. If we want to add extra functionality to it (i.e. track statistics),  we need to add it to the just one place.
2. Clean code: it is much easier to understand the code which has only one entry point for each operation.
3. Effectiveness: we send cancellation task to the all cluster nodes, but it might be excessive. The query can be finished on some nodes by the time of cancellation. Or query might be initially executed on a few nodes or a single node due to partition pruning. Only the initiator knows the actual query status and mapping, and therefore it should be responsible for the cancellation.

Reply:
> And this entry point should be located on the initiator.
> Convenience: single entry point is easier to control

The initiator is the reason for the issue.
Imagine a case when CPU usage on the initiator is 100% because it flooded with the query results. In that case, the initiator will handle cancel message very slowly that can lead to a whole cluster disaster. 

I think we shouldn't wait until the initiator can proceed cancel message.
We should cancel it as fast as we can.

> Effectiveness: we send cancellation task to all cluster nodes, but it might be excessive

We can try to reduce the number of nodes to the number of affinity nodes.
Anyway, cancellation of the query is not a performance-critical operation.

Reply:
Every node in the cluster can be overloaded, even the one where we are going to send the kill task from. This is not an issue.
We should have a single straightforward protocol for starting and cancelling the tasks. Existing protocol for cancelling tasks from the initiator is a good candidate for it. If instead we have a bunch of such protocols we will end up with a messy code and numerous subtle bugs which are not very easy to cope with. 
I think that proposed approach is not bug-free and there are possible races between task start and task cancellation messages. What would happen if cancel task request arrived before start task request? I am not sure this situation will be gracefully handled.
IMO we need to stick the approach with a single cancellation protocol from the initiator. It is solid and less erroneous solution.

Reply:
> Every node in the cluster can be overloaded, even the one where we are going to send the kill task from. 
> This is not an issue.

This is the issue I'm trying to solve with these changes.

> I think that proposed approach is not bug-free

Can you, please, clarify, what tests should be added to check the changes?

> What would happen if cancel task request arrived before start task request? 

Then we would get an error "Query not found".

Reply:
We will end up with memory and resources leak. Because the cancel task message arrives first and we get  "Query not found" here. But later the start task message arrives, the task starts execution but no one will cancel it because cancel task will not be resent.

> Then we would get an error "Query not found".

You may use custom communication SPI to delay start task messages and reorder them with cancel task messages.
> Can you, please, clarify, what tests should be added to check the changes?